codec/src/main/java/io/netty/handler/codec/ByteString.java

```diff
@@ -0,0 +1,756 @@
+     * will be shared. If this shared memory changes then {@link #arrayChanged()} must be called.
+     */
+    public ByteString(byte[] value, int start, int length, boolean copy) {
+        if (start < 0 || start > checkNotNull(value, "value").length - length) {
+            throw new IndexOutOfBoundsException("expected: " + "0 <= start(" + start + ") <= start + length(" + length
+                            + ") <= " + "value.length(" + value.length + ')');
+        }
+
+        if (copy || start != 0 || length != value.length) {
+            this.value = Arrays.copyOfRange(value, start, start + length);
+        } else {
+            this.value = value;
+        }
+    }
+
+    /**
+     * @see {@link #getBytes(ByteBuffer)}
+     */
+    public ByteString(ByteBuffer value) {
+        this.value = getBytes(value);
+    }
+
+    /**
+     * @see {@link #getBytes(ByteBuffer, int, int)}
+     */
+    public ByteString(ByteBuffer value, int start, int length) {
+        this.value = getBytes(value, start, length, true);
+    }
+
+    /**
+     * @see {@link #getBytes(ByteBuffer, int, int, boolean)}
+     */
+    public ByteString(ByteBuffer value, int start, int length, boolean copy) {
+        this.value = getBytes(value, start, length, copy);
+    }
+
+    /**
+     * @see {@link #getBytes(ByteBuf)}
+     */
+    public ByteString(ByteBuf buf) {
+        this.value = getBytes(buf);
+    }
+
+    /**
+     * @see {@link #getBytes(ByteBuf, int, int)}
+     */
+    public ByteString(ByteBuf buf, int start, int length) {
+        this.value = getBytes(buf, start, length, true);
+    }
+
+    /**
+     * @see {@link #getBytes(ByteBuf, int, int, boolean)}
+     */
+    public ByteString(ByteBuf buf, int start, int length, boolean copy) {
+        this.value = getBytes(buf, start, length, copy);
+    }
+
+    /**
+     * @see {@link #getBytes(char[], Charset)}
+     */
+    public ByteString(char[] value, Charset charset) {
+        this.value = getBytes(value, charset);
+    }
+
+    /**
+     * @see {@link #getBytes(char[], Charset, int, int)}
+     */
+    public ByteString(char[] value, Charset charset, int start, int length) {
+        this.value = getBytes(value, charset, start, length);
+    }
+
+    /**
+     * @see {@link #getBytes(CharSequence, Charset)}
+     */
+    public ByteString(CharSequence value, Charset charset) {
+        this.value = getBytes(value, charset);
+    }
+
+    /**
+     * @see {@link #getBytes(CharSequence, Charset, int, int)}
+     */
+    public ByteString(CharSequence value, Charset charset, int start, int length) {
+        this.value = getBytes(value, charset, start, length);
+    }
+
+    /**
+     * Create a copy of the underlying storage from {@link value} into a byte array.
+     * The copy will start at {@link ByteBuffer#position()} and copy {@link ByteBuffer#remaining()} bytes.
+     */
+    public static byte[] getBytes(ByteBuffer value) {
```

Comment:
I guess they don't have to be.  The thought was incase you don't want to have a `ByteString` (which Ideally should be unmodifiable) object you could directly to `byte[]`.